osbuild/solver/dnf.py

```diff
@@ -322,9 +337,78 @@ def depsolve(self, arguments):
+                    transaction.get("module-enable-specs", []),
+                ):
+                    # we don't particularly care about the NSVCAP here, just the module
+                    # packages that were previously selected
+                    module_packages, _ = self.base_module.get_modules(module_spec)
+
+                    # we now need to do an annoying dance as multiple modules could be
+                    # returned by `.get_modules`, we need to select the *same* one as
+                    # previously selected. we do this by checking if any of the module
+                    # packages are in the packages set marked for installation.
+
+                    # XXX this is a result of not being able to get the enabled modules
+                    # XXX from the transaction, if that turns out to be possible then
+                    # XXX we can get rid of these shenanigans
+                    for module_package in module_packages:
+                        module_nevras = module_package.getArtifacts()
+
+                        if any(module_nevra in package_nevras for module_nevra in module_nevras):
+                            # a package from this module is being installed so we must
+                            # use this module
+                            modules.append(module_package)
+
+                            # XXX we are probably able to skip the rest of the `module_packages`
+                            # XXX here if we want since no two modules can be enabled by the same
+                            # XXX name anyhow
+
+        # now we have the information we need about modules so we need to return *some*
+        # information to who is using the depsolver so they can use that information to
+        # enable these modules in the artifact
+
+        # there are two files that matter for each module that is used, the caller needs
+        # to write a file to `/etc/dnf/modules.d/{module_name}.module` to enable the
+        # module for dnf
+
+        # the caller also needs to set up `/var/lib/dnf/modulefailsafe/` with the contents
+        # of the modulemd for the selected modules, this is to ensure that even when a
+        # repository is disabled or disappears that non-modular content can't be installed
+        # see: https://dnf.readthedocs.io/en/latest/modularity.html#fail-safe-mechanisms
+
+        # XXX for each module we provide the contents for both files, this allows the
+        # XXX the caller to add inline sources and copy stages we *might* want to do this
+        # XXX differently instead by having specific stages and a specific format?
+        for module in modules:
+            response["modules"][module.getName()] = {
+                "module-file": textwrap.dedent(f"""\
+                    [{module.getName()}]
+                    name={module.getName()}
+                    stream={module.getStream()}
+                    profiles=common
```

Comment:
What if a module doesn't have a common profile? Like the postgresql module only has

``` 
...
  profiles:
    client:
      rpms:
      - postgresql
    server:
      rpms:
      - postgresql-server
...
```

Should we just add all profiles?

Reply:
Ah, good catch yes. We either parse the module specification and see if it contains a selected profile or we parse the module metadata and select the default profile (as `dnf` does).

We can't really select all profiles as that would lead to a mismatch of what `dnf` installed and what it thinks is installed.